e3nn/io/_cartesian_tensor.py

```diff
@@ -1,10 +1,64 @@
 from typing import Optional
+from threading import local
+import weakref
+import copy
 
 import torch
 
 from e3nn import o3
 
 
+# CartesianTensor shouldn't carry a torch.nn.Module child
+# but we don't want to rebuild the RTP every call
+# cache it here in thread local storage (to avoid
+# possibility of weird threading bugs)
+_RTP_CACHE = local()
+_RTP_CACHE.cache = weakref.WeakKeyDictionary()  # Dict[_formula, Dict[(device, dtype), ReducedTensorProducts]]
+
+
+# need something that gets garbage collected; raw str won't do
+class _formula:
```

Comment:
I think it would be ever so slightly cleaner to use a named tuple here, that way it's immutable.

`_formula = collections.namedtuple('_formula', 'formula')`

Reply:
I tried this but got errors— apparently you can't `weakref` a `NamedTuple`...

Reply:
although I tried the way where you subclass `typing.NamedTuple`. Maybe using `collections.namedtuple` would work? Not sure if they are synonymous

Reply:
ok they are the same: https://stackoverflow.com/questions/50766461/whats-the-difference-between-namedtuple-and-namedtuple

Reply:
Ah, bummer.  Ok, then all good as it is.